File: src/ldap_filter.rs
use ldap3::{ldap_escape};

pub fn true_() -> &'static str {
    "(objectClass=*)"
}

/*pub fn group() -> &'static str {
    "(objectClass=groupOfNames)"
}*/

pub fn eq(attr: &str, val: &str) -> String {
    format!("({}={})", attr, ldap_escape(val))
}

pub fn _not(filter: &str) -> String {
    format!("(!{})", filter)
}

pub fn and2(filter1: &str, filter2: &str) -> String {
    format!("(&{}{})", filter1, filter2)
}

pub fn or(l : Vec<String>) -> String {
    match &l[..] {
        [filter] => filter.to_owned(),
        _ => format!("(|{})", l.concat()),
    }
}

pub fn rdn(rdn: &str) -> String {
    format!("({})", rdn)
}

pub fn member(dn: &str) -> String {
    eq("member", dn)
}

pub fn sgroup_self_and_children(id: &str) -> String {
    let id = ldap_escape(id);
    format!("(cn={}*)", &id)
}

pub fn sgroup_children(id: &str) -> String {
    if id.is_empty() {
        "(cn=*)".to_owned()
    } else {
        let id = ldap_escape(id);
        format!("(&(cn={}*)(!(cn={})))", &id, &id)
    }
}

File: server/ldap_filter.ts
import { escape } from 'ldapjs/lib/filters/escape';
import { Dn, Option } from './my_types';

const and = (filters: string[]) => {
    // simplify empty AND/OR which are not handled by ldapjs.parseFilter
    const l = filters.filter(e => e !== '(&)')
    return l.length === 1 ? l[0] : l.includes('(|)') ? '(|)' : "(&" + l.join('') + ")"
}

const or = (filters : string[]) => {
    // simplify empty AND/OR which are not handled by ldapjs.parseFilter
    const l = filters.filter(e => e !== '(|)')
    return l.length === 1 ? l[0] : l.includes('(&)') ? '(&)' : "(|" + l.join('') + ")"
}

const eq = (attr: string, val: string) => (
    `(${attr}=${escape(val)})`
)

export default {
    or,
    and,
    eq,

    true_: () => (
        "(objectClass=*)"
    ),

    presence: (attr: string) => (
        `(${attr}=*)`
    ),

        
    not: (filter: string) => (
        `(!${filter})`
    ),
            
    and2: (filter1: string, filter2: string) => (
        and([filter1, filter2])
    ),
    and2_if_some: (filter1: string, filter2: Option<string>) => (
        filter2 ? and([filter1, filter2]) : filter1
    ),

    rdn: (rdn: string) => (
        `(${rdn})`
    ),

    member: (dn: Dn) => (
        eq("member", dn)
    ),

    sgroup_self_and_children: (id: string) => (
        `(cn=${escape(id)}*)`
    ),

    sgroup_children: (id: string) => {
        if (id === '') {
            return "(cn=*)"
        } else {
            id = escape(id);
            return `(&(cn=${id}*)(!(cn=${id})))`
        }
    },

}